### PR TITLE
Add legend face option

### LINKED ISSUE #1487: Feature Request: change legend symbol/label to other shapes instead of triangle only
`legend_elements = [[str(unique_label), cmap.colors[i]] for i, unique_label in enumerate(unique_labels)]`
`p.add_legend(legend_elements)`
Capture

Hi,
It will be good if we can have a feature to change the triangle to some other shapes. e.g. a circle or a rectangle with the right side numbers written within it.

Bests,
Kai.

### PR DESCRIPTION
**Pull request by tkoyama010**
### Overview

Add legend face option.

 

resolves #1487

### Details

```python
import pyvista

pl = pyvista.Plotter()
legend_elements = [[str(i), (1, 1, 1)] for i in range(13)]
pl.add_legend(legend_elements, size = [1.0, 1.0], origin = [0.0, 0.0])
pl.show(screenshot="legend_triangle_face.png")
```

```python
pl = pyvista.Plotter()
legend_elements = [[str(i), (1, 1, 1)] for i in range(13)]
pl.add_legend(legend_elements, size = [1.0, 1.0], origin = [0.0, 0.0], face = "circle")
pl.show(screenshot="legend_triangle_circle.png")
```

```python
pl = pyvista.Plotter()
legend_elements = [[str(i), (1, 1, 1)] for i in range(13)]
pl.add_legend(legend_elements, size = [1.0, 1.0], origin = [0.0, 0.0], face = "rectangle")
pl.show(screenshot="legend_triangle_rectangle.png")
```

### CODE REVIEW

**Review by akaszynski — COMMENTED** [review]
@tkoyama010, I like what you did here, but I think that `single_circle` and the other methods should be private.  Our top level namespace is already quite busy, do you think we could make a "geometry" module that includes all our basic geometry methods in a later PR?  Then we could follow our existing pattern of `Circle`, `Triangle`, `Rectangle` with:

```
import pyvista
tri = pyvista.geometry.Triangle()
```

Let's shoot for that.

**Comment by tkoyama010** [comment]
@akaszynski That's a good idea. But I think it is better to include [`geometric_objects`](https://github.com/pyvista/pyvista/blob/feat/legend_face_option/pyvista/utilities/geometric_objects.py) module. How do you think?

**Comment by akaszynski** [comment]
Perfect, good idea.

File: pyvista/utilities/geometric_objects.py [hunk 1]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
     ug.InsertNextCell(pyramid.GetCellType(), pyramid.GetPointIds())
 
     return pyvista.wrap(ug)
+
+
+def Triangle(points = [[0, 0, 0], [1, 0, 0], [0.5, 0.707, 0]]):
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:994]:
```suggestion
def Triangle(points=[[0, 0, 0], [1, 0, 0], [0.5, 0.707, 0]]):
```

File: pyvista/utilities/geometric_objects.py [hunk 2]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+
+    Parameters
+    ----------
+    points : np.ndarray or list
+        Points of the triangle.
+
+    Returns
+    -------
+    pyvista.PolyData
+        triangle mesh.
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1005]:
```suggestion
        Triangle mesh.
```

File: pyvista/utilities/geometric_objects.py [hunk 3]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+    check_valid_vector(points[1], 'points[1]')
+    check_valid_vector(points[2], 'points[2]')
+
+    cells = np.array([[3, 0, 1, 2]], ctypes.c_long)
+    triangle = pyvista.PolyData(points, cells)
+
+    return triangle
+
+
+def Rectangle(points = [[1, 0, 0], [1, 1, 0], [0, 1, 0], [0, 0, 0]]):
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1029]:
```suggestion
def Rectangle(points=[[1, 0, 0], [1, 1, 0], [0, 1, 0], [0, 0, 0]]):
```

File: pyvista/utilities/geometric_objects.py [hunk 4]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+
+    Parameters
+    ----------
+    points : np.ndarray or list
+        Points of the rectangle.
+
+    Returns
+    -------
+    pyvista.PolyData
+        rectangle mesh.
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1040]:
```suggestion
        Rectangle mesh.
```

File: pyvista/utilities/geometric_objects.py [hunk 5]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+    check_valid_vector(points[2], 'points[2]')
+    check_valid_vector(points[3], 'points[3]')
+
+    cells = np.array([[4, 0, 1, 2, 3]], ctypes.c_long)
+    rectangle = pyvista.PolyData(points, cells)
+
+    return rectangle
+
+
+def Circle(radius = 0.5):
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1066]:
```suggestion
def Circle(radius=0.5):
```

File: pyvista/utilities/geometric_objects.py [hunk 6]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+    """Create a single PolyData circle defined by radius.
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1067]:
```suggestion
    """Create a single PolyData circle defined by radius in the XY plane.
```

File: pyvista/utilities/geometric_objects.py [hunk 7]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+        circle mesh.
+
+    Examples
+    --------
+    >>> import pyvista
+    >>> radius = 0.5
+    >>> circle = pyvista.Circle(radius)
+    >>> circle.plot(show_edges=True, line_width=5)
+    """
+    resolution = 100
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1086]:
```suggestion
```

File: pyvista/utilities/geometric_objects.py [hunk 8]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+
+
+def Circle(radius = 0.5):
+    """Create a single PolyData circle defined by radius.
+
+    Parameters
+    ----------
+    radius : float, optional
+        Radius of circle.
+
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1073]:
```suggestion
    radius : float, optional
        Radius of circle.

    resolution : int, optional
        Number of points on the circle.

```

File: pyvista/utilities/geometric_objects.py [hunk 9]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+    >>> radius = 0.5
+    >>> circle = pyvista.Circle(radius)
+    >>> circle.plot(show_edges=True, line_width=5)
+    """
+    resolution = 100
+    points = np.zeros((resolution, 3))
+    theta = np.linspace(0.0, 2.0*np.pi, resolution)
+    points[:, 0] = radius * np.cos(theta)
+    points[:, 1] = radius * np.sin(theta)
+    cells = np.array([np.append(np.array([resolution]), np.arange(100))], ctypes.c_long)
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1091]:
```suggestion
    cells = np.array([np.append(np.array([resolution]), np.arange(resolution))], ctypes.c_long)
```

File: pyvista/plotting/plotting.py [hunk 10]
```diff
@@ -3466,7 +3476,19 @@ def add_legend(self, labels=None, bcolor=(0.5, 0.5, 0.5), border=False,
...
         else:
             self.legend.SetNumberOfEntries(len(labels))
-            legendface = pyvista.single_triangle()
+
+            if (face is None) or (face == "triangle"):
+                legendface = pyvista.Triangle()
+            elif (face == "circle"):
+                legendface = pyvista.Circle()
+            elif (face == "rectangle"):
+                legendface = pyvista.Rectangle()
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:3485]:
```suggestion
            if face is None or face == "triangle":
                legendface = pyvista.Triangle()
            elif face == "circle":
                legendface = pyvista.Circle()
            elif face == "rectangle":
                legendface = pyvista.Rectangle()
```

File: pyvista/plotting/plotting.py [hunk 11]
```diff
@@ -3466,7 +3476,19 @@ def add_legend(self, labels=None, bcolor=(0.5, 0.5, 0.5), border=False,
...
+            else:
+                raise ValueError('Invalid face.  Must be one of the following:\n'
```
Comment (akaszynski) [inline pyvista/plotting/plotting.py:3487]:
```suggestion
                raise ValueError(f'Invalid face "{face}".  Must be one of the following:\n'
```

File: pyvista/utilities/geometric_objects.py [hunk 12]
```diff
@@ -987,3 +989,106 @@ def Pyramid(points):
...
+
+    Parameters
+    ----------
+    radius : float, optional
+        Radius of circle.
+
+    Returns
+    -------
+    pyvista.PolyData
+        circle mesh.
```
Comment (akaszynski) [inline pyvista/utilities/geometric_objects.py:1077]:
```suggestion
        Circle mesh.
```

**Review by akaszynski — APPROVED** [review]
Nice work on this @tkoyama010!

**Pull request merged**